lib/Dialect/Ibis/Transforms/IbisPortrefLowering.cpp

```diff
@@ -0,0 +1,337 @@
+              "port reference");
+
+    // Replace the inner portref + port access with a "raw" port.
+    rewriter.setInsertionPoint(op);
+    if (d == Direction::Input) {
+      // references to inputs becomes outputs (write from this container)
+      auto rawOutput = rewriter.create<OutputPortOp>(
+          op.getLoc(), op.getPortName(), innerType);
+
+      // Replace writes to the unwrapped port with writes to the new port.
+      for (auto *unwrappedPortUser :
+           llvm::make_early_inc_range(portUnwrapper.getResult().getUsers())) {
+        PortWriteOp portWriter = dyn_cast<PortWriteOp>(unwrappedPortUser);
+        if (!portWriter || portWriter.getPort() != portUnwrapper.getResult())
+          continue;
+
+        // Replace the source port of the write op with the new port.
+        rewriter.replaceOpWithNewOp<PortWriteOp>(portWriter, rawOutput,
+                                                 portWriter.getValue());
+      }
+    } else {
+      // References to outputs becomes inputs (read from this container)
+      auto rawInput = rewriter.create<InputPortOp>(op.getLoc(),
+                                                   op.getPortName(), innerType);
+      rewriter.replaceAllUsesWith(portUnwrapper.getResult(), rawInput);
+
+      // Replace all ibis.port.read ops with a read of the new input.
+      for (auto *portUser :
+           llvm::make_early_inc_range(portUnwrapper.getResult().getUsers())) {
+        PortReadOp portReader = dyn_cast<PortReadOp>(portUser);
+        if (!portReader || portReader.getPort() != portUnwrapper.getResult())
+          continue;
+
+        rewriter.replaceOpWithNewOp<PortReadOp>(portReader, rawInput);
+      }
+    }
+
+    // Finally, remove the port unwrapper and the original input port.
+    rewriter.eraseOp(portUnwrapper);
+    rewriter.eraseOp(op);
+
+    return success();
+  }
+};
+
+class OutputPortConversionPattern : public PortLoweringPattern<OutputPortOp> {
+public:
+  using PortLoweringPattern::PortLoweringPattern;
+  using OpAdaptor = typename PortLoweringPattern<OutputPortOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(OutputPortOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    PortRefType innerPortRefType = op.getType().cast<PortRefType>();
+    Type innerType = innerPortRefType.getPortType();
+    Direction d = innerPortRefType.getDirection();
+
+    // Locate the portwrapper - this is a writeOp with the output portref as
+    // the portref value.
+    PortWriteOp portWrapper;
+    for (auto *user : op.getResult().getUsers()) {
+      portWrapper = dyn_cast<PortWriteOp>(user);
+      if (portWrapper && portWrapper.getPort() == op.getResult())
+        break;
+    }
+
+    if (!portWrapper)
+      return rewriter.notifyMatchFailure(
+          op, "expected an ibis.port.write to wrap the output portref");
+
+    rewriter.setInsertionPoint(op);
```

Comment:
Should there be an insertion guard here?

Reply:
Wouldn't hurt to have one.